Restrict formatting controls in write mode (contentOnly)

getdave opened the pull request:
## What

Restricts RichText formatting controls when in write mode (contentOnly) to provide a simpler, more focused editing experience while maintaining essential content functionality. Introduces a private `essentialFormatKey` Symbol for format registration to distinguish between essential and non-essential formats.

## Why

Write mode should be as simple as possible to help users focus on content creation without distractions from advanced formatting options. Currently, users see all formatting controls including strikethrough, highlight, and other advanced options even in write mode. However, links are essential content functionality and should remain available in Write Mode, especially for Navigation blocks.

## How

- **Write Mode detection in RichText component** - Added Write Mode detection with experiment flag check in the main RichText component
- **Centralized format filtering** - Enhanced `useFormatTypes` hook to filter formats based on Write Mode and essential status
- **Single store subscription** - Uses one centralized subscription for Write Mode detection instead of multiple per format component
- **Introduced private `essentialFormatKey` Symbol** for format registration (using WordPress Core's private API pattern)
- **By default, all formats are non-essential** except for hardcoded essential formats: **bold**, **italic**, and **link**
- **Third-party formats are non-essential by default** to prevent them from cluttering the Write Mode toolbar
- Formats can opt-in to be essential by setting `[ essentialFormatKey ]: true` in their registration
- **Only affects Write Mode when experiment is active** - early return when `window.__experimentalEditorWriteMode` is false ensures minimal overhead
- In write mode: Shows only essential formats (bold, italic, link + any formats marked as essential)
- In regular mode: Shows full formatting toolbar (all available formats)

## Testing

1. Enable Write Mode experiment in Gutenberg > Experiments
2. Switch to write mode in the editor
3. Select text in a RichText block (paragraph, heading, etc.)
4. Verify bold, italic, and link buttons are visible in the inline formatting toolbar
5. Verify non-essential formats (strikethrough, highlight, etc.) are hidden
6. Test link functionality in Write Mode works correctly
7. Switch back to regular mode and verify full formatting toolbar reappears
8. Verify performance is optimal when Write Mode experiment is disabled

## Testing Format Registration

You can test the essential format registration by applying the following diff to make strikethrough essential:

```diff
--- a/packages/format-library/src/strikethrough/index.js
+++ b/packages/format-library/src/strikethrough/index.js
@@ -1,6 +1,7 @@
 /**
  * WordPress dependencies
  */
+import { unlock } from '../lock-unlock';
 import { __ } from '@wordpress/i18n';
 import { toggleFormat } from '@wordpress/rich-text';
 import {
@@ -8,6 +9,9 @@ import {
 	RichTextShortcut,
 	__unstableRichTextInputEvent,
 	privateApis as blockEditorPrivateApis,
 } from '@wordpress/block-editor';
 import { formatStrikethrough } from '@wordpress/icons';
 
+const { essentialFormatKey } = unlock( blockEditorPrivateApis );
+
 const name = 'core/strikethrough';
 const title = __( 'Strikethrough' );
 
@@ -15,6 +19,7 @@ export const strikethrough = {
 	name,
 	title,
 	tagName: 's',
+	[ essentialFormatKey ]: true,
 	className: null,
 	edit( { isActive, value, onChange, onFocus } ) {
```

After applying this diff:
1. Build the packages if needed
2. Enable Write Mode experiment
3. Switch to Write Mode
4. Verify that strikethrough now appears in the formatting toolbar (in the dropdown).
5. Verify that other non-essential formats remain hidden

## Screenshots
<img width="1098" height="684" alt="Screen Shot 2025-08-07 at 11 40 23" src="https://github.com/user-attachments/assets/dfd86479-b3bf-4c18-8904-88a9ee26e54e" />



## Breaking Changes

None - this is a UI enhancement that maintains all existing functionality in regular editing mode.

## Performance Considerations

- **Zero overhead when Write Mode experiment is disabled** - Filtering logic is bypassed entirely
- **Single store subscription** - Uses one centralized subscription for Write Mode detection
- **Centralized filtering** - All format filtering logic is handled in `useFormatTypes` hook
- **Efficient memoization** - Uses React's built-in memoization for optimal re-rendering
- **Early return optimization** - Bypasses filtering logic when experiment is disabled

## API Considerations

- `essentialFormatKey` is a private Symbol following WordPress Core's private API pattern
- Third-party formats are non-essential by default to prevent toolbar clutter
- Formats can opt-in to be essential by using the private Symbol in their registration
- Private API ensures no public API commitment while allowing Core packages to use the feature

## Related

This change supports the Navigation Block Write Mode improvements (see #70977) where navigation blocks need to be editable in Write Mode with link functionality available.

getdave committed:
Restrict formatting controls in write mode (contentOnly)

- Add useBlockEditingMode() hook to detect write mode
- In contentOnly mode, only show bold and italic controls
- Remove link and unknown formats from write mode toolbar
- Hide 'More' dropdown with additional formatting options in write mode
- Keep write mode simple and focused on essential content editing

Mamaduka commented:
The content-only editing can be used without the write mode. Do we want this change to affect blocks inside containers when with `templateLock: 'contentOnly'`? @jeryj and I discovered the same problem in the recent PR - #70946.

Hardcoding "essential" formats seems limiting for extenders. What if they provide custom formats that are essential for their editorial teams?

Maybe we should consider letting formats declare themself as essential for write mode.

## Screenshot
<img width="1349" height="702" alt="CleanShot 2025-08-05 at 14 01 01" src="https://github.com/user-attachments/assets/63152840-0cb7-4473-82a8-41b6e64c27e6" />

getdave commented:
@Mamaduka Thanks for flagging. It's been a while.

So is Write Mode equivalent to `isNavigationMode && blockEditingMode === 'contentOnly'`?

getdave committed:
Update format toolbar to use correct Write Mode pattern

Mamaduka commented:
> I agree with this in principle. I do have concerns over it being abused. If I'm a plugin author trying to promote my tooling, why would I not define it as essential?

IMO, all public APIs can be abused; the best we can do here is to tuck them in a dropdown (similar to the current system).

P.S. It is still possible for custom format libraries to render items in the write mode toolbar if they know how SlotFills work.

getdave committed:
Add write mode format filtering with essential flag

- Add withWriteModeFilter HOC to filter format edit components in write mode
- Add essential property to format registrations (default: false)
- Mark core formats (bold, italic, link, strikethrough) as essential
- Apply HOC in FormatEdit component to filter non-essential formats
- Maintain backward compatibility for existing formats
- Enable third-party plugins to opt formats into write mode via essential flag

This allows write mode to show only essential formatting controls while
maintaining full functionality in default mode.

getdave committed:
Mark strikethrough as non-essential again

getdave committed:
Restore changes to format edit

getdave committed:
Tidy changes to toolbar

getdave committed:
Restore toolbar to original

getdave commented on packages/block-editor/src/components/rich-text/with-write-mode-filter.js:
```diff
@@ -0,0 +1,38 @@
+		const isWriteMode = isNavigationMode && isContentOnlyMode;
+
+		// In write mode, only show essential formats
+		if ( isWriteMode && ! formatSettings?.essential ) {
```
Note that if `essential` is not defined then the default behaviour is not to show in Write Mode.

getdave committed:
Add perf opt

getdave committed:
Optimize withWriteModeFilter HOC performance

- Add early return for Write Mode experiment flag check
- Avoid applying HOC when experiment is not enabled
- Maintain existing optimizations for essential formats
- Improve performance by reducing unnecessary wrapper components

getdave committed:
Use unstable

getdave committed:
Revert unneeded changes to formats

getdave commented:
> Maybe we should consider letting formats declare themself as essential for write mode.

@Mamaduka @jeryj I went ahead and implemented one route for this. I didn't want to remove formats entirely or stop their registration because that's not the goal of Write Mode. All we want to do is not provide the _UI_ in write mode.

I found the best way to do that was to wrap the format's `Edit` definition with a HoC that will only render in Write Mode if the format is registered with `__unstableEssential` (feel free to suggest alternative names).

I provided early escape hatches and also ensured it doesn't filter the component if the experiment isn't active.

I updated the description and you can follow the testing instructions there to get a feel for how this works.

The result is the same as per the original implementation, but just with the benefit that extenders can choose to mark their formats as essential to force them to show in Write Mode (under the dropdown).

jeryj commented:
What if `packages/block-editor/src/components/rich-text/with-write-mode-filter.js` or similar was utilized within the `<BlockToolbar />` component instead? Then any Block Toolbar button or item could add the `_unstableEssential` (or whatever we end up calling it).

Mamaduka commented:
@getdave, what do you think about handling all related filtering inside the `useFormatTypes`? It already has similar logic for interactive formats. A small bonus point is that it won't create a store subscription for each format when a block is selected.

@jeryj, that might be more on a higher level, but it would be hard to control what `BlockToolbar` renders, since slots usually don't have that info.

<details><summary>Example patch, still requires full removal of `withWriteModeFilter`</summary>
<p>

```diff
diff --git a/packages/block-editor/src/components/rich-text/format-edit.js b/packages/block-editor/src/components/rich-text/format-edit.js
index 20d13c1d3f..a70b9f8f77 100644
--- a/packages/block-editor/src/components/rich-text/format-edit.js
+++ b/packages/block-editor/src/components/rich-text/format-edit.js
@@ -8,7 +8,6 @@ import { useContext, useMemo } from '@wordpress/element';
  * Internal dependencies
  */
 import BlockContext from '../block-context';
-import withWriteModeFilter from './with-write-mode-filter';
 
 const DEFAULT_BLOCK_CONTEXT = {};
 
@@ -38,9 +37,6 @@ function Edit( { onChange, onFocus, value, forwardedRef, settings } ) {
 		return null;
 	}
 
-	// Apply the write mode filter HOC
-	const FilteredEditFunction = withWriteModeFilter( EditFunction, settings );
-
 	const activeFormat = getActiveFormat( value, name );
 	const isActive = activeFormat !== undefined;
 	const activeObject = getActiveObject( value );
@@ -48,7 +44,7 @@ function Edit( { onChange, onFocus, value, forwardedRef, settings } ) {
 		activeObject !== undefined && activeObject.type === name;
 
 	return (
-		<FilteredEditFunction
+		<EditFunction
 			key={ name }
 			isActive={ isActive }
 			activeAttributes={ isActive ? activeFormat.attributes || {} : {} }
diff --git a/packages/block-editor/src/components/rich-text/index.js b/packages/block-editor/src/components/rich-text/index.js
index 768ffbb0cd..bd831558f7 100644
--- a/packages/block-editor/src/components/rich-text/index.js
+++ b/packages/block-editor/src/components/rich-text/index.js
@@ -132,8 +132,12 @@ export function RichTextWrapper(
 			return { isSelected: false };
 		}
 
-		const { getSelectionStart, getSelectionEnd } =
-			select( blockEditorStore );
+		const {
+			getSelectionStart,
+			getSelectionEnd,
+			getBlockEditingMode,
+			isNavigationMode,
+		} = select( blockEditorStore );
 		const selectionStart = getSelectionStart();
 		const selectionEnd = getSelectionEnd();
 
@@ -154,15 +158,21 @@ export function RichTextWrapper(
 			selectionStart: isSelected ? selectionStart.offset : undefined,
 			selectionEnd: isSelected ? selectionEnd.offset : undefined,
 			isSelected,
+			isWriteMode:
+				isNavigationMode() &&
+				getBlockEditingMode( clientId ) === 'contentOnly',
 		};
 	};
-	const { selectionStart, selectionEnd, isSelected } = useSelect( selector, [
-		clientId,
-		identifier,
-		instanceId,
-		originalIsSelected,
-		isBlockSelected,
-	] );
+	const { selectionStart, selectionEnd, isSelected, isWriteMode } = useSelect(
+		selector,
+		[
+			clientId,
+			identifier,
+			instanceId,
+			originalIsSelected,
+			isBlockSelected,
+		]
+	);
 
 	const { disableBoundBlock, bindingsPlaceholder, bindingsLabel } = useSelect(
 		( select ) => {
@@ -323,8 +333,9 @@ export function RichTextWrapper(
 	} = useFormatTypes( {
 		clientId,
 		identifier,
-		withoutInteractiveFormatting,
 		allowedFormats: adjustedAllowedFormats,
+		withoutInteractiveFormatting,
+		disableNoneEssentialFormatting: isWriteMode,
 	} );
 
 	function addEditorOnlyFormats( value ) {
diff --git a/packages/block-editor/src/components/rich-text/use-format-types.js b/packages/block-editor/src/components/rich-text/use-format-types.js
index 0bbebbf262..fcb58a852c 100644
--- a/packages/block-editor/src/components/rich-text/use-format-types.js
+++ b/packages/block-editor/src/components/rich-text/use-format-types.js
@@ -56,35 +56,48 @@ function getPrefixedSelectKeys( selected, prefix ) {
  * This hook provides RichText with the `formatTypes` and its derived props from
  * experimental format type settings.
  *
- * @param {Object}  $0                              Options
- * @param {string}  $0.clientId                     Block client ID.
- * @param {string}  $0.identifier                   Block attribute.
- * @param {boolean} $0.withoutInteractiveFormatting Whether to clean the interactive formatting or not.
- * @param {Array}   $0.allowedFormats               Allowed formats
+ * @param {Object}  options                                Options
+ * @param {string}  options.clientId                       Block client ID.
+ * @param {string}  options.identifier                     Block attribute.
+ * @param {Array}   options.allowedFormats                 Allowed formats
+ * @param {boolean} options.withoutInteractiveFormatting   Whether to clean the interactive formatting or not.
+ * @param {boolean} options.disableNoneEssentialFormatting Whether to disable none-essential formatting or not.
  */
 export function useFormatTypes( {
 	clientId,
 	identifier,
-	withoutInteractiveFormatting,
 	allowedFormats,
+	withoutInteractiveFormatting,
+	disableNoneEssentialFormatting = false,
 } ) {
 	const allFormatTypes = useSelect( formatTypesSelector, [] );
 	const formatTypes = useMemo( () => {
-		return allFormatTypes.filter( ( { name, interactive, tagName } ) => {
-			if ( allowedFormats && ! allowedFormats.includes( name ) ) {
-				return false;
-			}
+		return allFormatTypes.filter(
+			( { name, interactive, tagName, __unstableEssential } ) => {
+				if ( allowedFormats && ! allowedFormats.includes( name ) ) {
+					return false;
+				}
 
-			if (
-				withoutInteractiveFormatting &&
-				( interactive || interactiveContentTags.has( tagName ) )
-			) {
-				return false;
-			}
+				if ( disableNoneEssentialFormatting && ! __unstableEssential ) {
+					return false;
+				}
+
+				if (
+					withoutInteractiveFormatting &&
+					( interactive || interactiveContentTags.has( tagName ) )
+				) {
+					return false;
+				}
 
-			return true;
-		} );
-	}, [ allFormatTypes, allowedFormats, withoutInteractiveFormatting ] );
+				return true;
+			}
+		);
+	}, [
+		allFormatTypes,
+		allowedFormats,
+		disableNoneEssentialFormatting,
+		withoutInteractiveFormatting,
+	] );
 	const keyedSelected = useSelect(
 		( select ) =>
 			formatTypes.reduce( ( accumulator, type ) => {

``` 

</p>
</details>

Mamaduka commented on packages/format-library/src/bold/index.js:
```diff
@@ -18,6 +18,7 @@ export const bold = {
 	title,
 	tagName: 'strong',
 	className: null,
+	__unstableEssential: true,
```
Core should no longer ship public APIs with `__unstable|experimental` prefixes. We should either make it private or stable.

getdave replied:
Thanks for flagging.

I've gone ahead and used a `Symbol` as a private api constant in place of the hardcoded key.

This means it's only available internally for now. 

If Write Mode makes it into a release then we can stablise the prop. If not we can safely remove it on the basis that it's private.

Is that suitable?

Mamaduka replied:
I think that's a good approach. Thanks, Dave!

getdave committed:
Refactor Write Mode format filtering to use useFormatTypes hook

- Remove withWriteModeFilter HOC approach in favor of centralized filtering
- Add Write Mode detection to RichText component with experiment flag check
- Enhance useFormatTypes hook with disableNoneEssentialFormatting parameter
- Centralize all format filtering logic in one place for better performance
- Eliminate individual store subscriptions per format component
- Maintain same functionality with improved architecture and performance

getdave commented:
> ...what do you think about handling all related filtering inside the `useFormatTypes`? 

I looked at this yesterday and dismissed it, but looking again it seems ideal. I'll make some adjustments.

Mamaduka commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -154,15 +158,22 @@ export function RichTextWrapper(
 			selectionEnd: isSelected ? selectionEnd.offset : undefined,
 			isSelected,
+			isWriteMode:
+				window?.__experimentalEditorWriteMode &&
+				isNavigationMode() &&
```
I think the `isNavigationMode` already checks the `window?.__experimentalEditorWriteMode`.

getdave committed:
Remove redundant experiment flag check in Write Mode detection

- isNavigationMode() already checks window.__experimentalEditorWriteMode internally
- Removes duplicate experiment flag check for cleaner, more efficient code
- Maintains same functionality while following existing codebase patterns

getdave committed:
Replace __unstableEssential with private Symbol essentialFormatKey

- Add essentialFormatKey to private-keys.js for internal Core usage
- Expose essentialFormatKey in private-apis.js for Core packages
- Update use-format-types.js to use private Symbol instead of __unstableEssential
- Update format library packages (bold, italic, link) to use private Symbol
- Follow correct import pattern: unlock from lock-unlock.js + privateApis from block-editor
- Remove unstable API prefix as per Core guidelines

getdave committed:
Improve variable naming: rename isWriteMode to isContentOnlyWriteMode

- More accurately reflects that we're checking both Write Mode and content-only editing
- Makes the dual condition check explicit: Write Mode + content-only block editing
- Improves code readability and maintainability
- Clarifies the relationship between Write Mode and content editing mode

Mamaduka reviewed (approved):
Thanks for the follow-ups, @getdave!

The changes work as advertised.

scruffian commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -154,15 +158,19 @@ export function RichTextWrapper(
-		originalIsSelected,
-		isBlockSelected,
-	] );
+	const { selectionStart, selectionEnd, isSelected, isContentOnlyWriteMode } =
```
I know you didn't introduce it in this PR, just wondering if there's a reason why `selector` is passed as a `const` here rather than just doing it inline like we usually do?

getdave replied:
Not sure. Seems odd.

scruffian reviewed (approved):
Looking great, thank you!

(I had to add `privateApis as blockEditorPrivateApis,` to the `@wordpress/block-editor` import, to get strikethrough to show up)

getdave commented:
## WP 6.9 note - Stabilization Needed

The `essential` property should be stabilized as a public API to allow plugin developers to create formats that are visible in Write Mode. 

This would enable plugins to provide essential formatting options that integrate seamlessly with the Write Mode experience.

Note that this PR implements the feature as private via a Symbol.

talldan commented:
> I'm thinking that actually we don't need the "is write mode" check at all as blocks marked as contentOnly shiould have consistent behaiour regardless of the wider editor mode.

You might need to check whether a block is a `section` as well, as I think they might also be content only. Not 100% sure, just trying to remember.

t-hamano commented:
This private API was removed in #75382, so the Dev Note is no longer required.
